File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/SerializationUtils.java
package com.thewizrd.shared_resources.utils;

import android.os.Build;

import java.nio.ByteBuffer;
import java.nio.charset.Charset;

public class SerializationUtils {
    public static byte[] booleanToBytes(boolean value) {
        return new byte[]{(byte) (value ? 1 : 0)};
    }

    public static boolean bytesToBool(byte[] data) {
        return !(data[0] == 0);
    }

    public static byte[] stringToBytes(String value) {
        return value.getBytes(Charset.forName("UTF-8"));
    }

    public static String bytesToString(byte[] data) {
        return new String(data);
    }

    public static byte[] intToBytes(int value) {
        ByteBuffer buf;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            buf = ByteBuffer.allocate(Integer.BYTES);
        } else {
            buf = ByteBuffer.allocate(4);
        }
        return buf.putInt(value).array();
    }

    public static int bytesToInt(byte[] data) {
        return ByteBuffer.wrap(data).getInt();
    }

    public static byte[] longToBytes(long value) {
        ByteBuffer buf;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            buf = ByteBuffer.allocate(Long.BYTES);
        } else {
            buf = ByteBuffer.allocate(8);
        }
        return buf.putLong(value).array();
    }

    public static long bytesToLong(byte[] data) {
        return ByteBuffer.wrap(data).getLong();
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/SerializationUtils.kt
@file:JvmName("SerializationUtils")

package com.thewizrd.shared_resources.utils

import java.nio.ByteBuffer
import java.nio.charset.Charset

fun Boolean.booleanToBytes(): ByteArray {
    return byteArrayOf((if (this) 1 else 0).toByte())
}

fun ByteArray.bytesToBool(): Boolean {
    return this[0] != 0.toByte()
}

fun String.stringToBytes(): ByteArray {
    return this.toByteArray(Charset.forName("UTF-8"))
}

fun ByteArray.bytesToString(): String {
    return String(this)
}

fun Int.intToBytes(): ByteArray {
    val buf = ByteBuffer.allocate(Int.SIZE_BYTES)
    return buf.putInt(this).array()
}

fun ByteArray.bytesToInt(): Int {
    return ByteBuffer.wrap(this).int
}

fun Long.longToBytes(): ByteArray {
    val buf = ByteBuffer.allocate(Long.SIZE_BYTES)
    return buf.putLong(this).array()
}

fun ByteArray.bytesToLong(): Long {
    return ByteBuffer.wrap(this).long
}

fun Char.charToBytes(): ByteArray {
    val buf = ByteBuffer.allocate(Char.SIZE_BYTES)
    return buf.putChar(this).array()
}

fun ByteArray.bytesToChar(): Char {
    return ByteBuffer.wrap(this).char
}